Libplanet/Net/AppProtocolVersion.cs

```diff
@@ -0,0 +1,167 @@
+        /// </summary>
+        public readonly IValue Extra;
+
+        /// <summary>
+        /// A signature which verifies <seealso cref="Signer"/>'s claim of a version.
+        /// </summary>
+        public readonly ImmutableArray<byte> Signature;
+
+        /// <summary>
+        /// A signer who claims presence of a version.
+        /// </summary>
+        public readonly Address Signer;
+
+        private static readonly Codec _codec = new Codec();
+
+        /// <summary>
+        /// Initializes an <see cref="AppProtocolVersion"/> value with field values.
+        /// </summary>
+        /// <param name="version">Sets the <see cref="Version"/>.</param>
+        /// <param name="extra">Sets the <see cref="Extra"/>.</param>
+        /// <param name="signature">Sets the <see cref="Signature"/>.</param>
+        /// <param name="signer">Gets the <see cref="Signer"/>.</param>
+        public AppProtocolVersion(
+            int version,
+            IValue extra,
+            ImmutableArray<byte> signature,
+            Address signer
+        )
+        {
+            Version = version;
+            Extra = extra;
+            Signature = signature;
+            Signer = signer;
+        }
+
+        /// <summary>
+        /// Claim a <paramref name="version"/> with <paramref name="extra"/> data and sign it
+        /// using the given private key.
+        /// </summary>
+        /// <param name="signer">A private key to sign the claim.</param>
+        /// <param name="version">A version to claim.</param>
+        /// <param name="extra">Extra data to claim.</param>
+        public AppProtocolVersion(PrivateKey signer, int version, IValue extra = null)
+            : this(
+                version,
+                extra,
+                ImmutableArray<byte>.Empty,
+                new Address((signer ?? throw new ArgumentNullException(nameof(signer))).PublicKey)
+            )
+        {
+            Signature = ImmutableArray.Create(signer.Sign(GetMessage()));
+        }
+
+        [Pure]
+        public static bool operator ==(AppProtocolVersion left, AppProtocolVersion right) =>
+            left.Equals(right);
+
+        [Pure]
+        public static bool operator !=(AppProtocolVersion left, AppProtocolVersion right) =>
+            !(left == right);
+
+        /// <summary>
+        /// Verifies whether the claim is certainly signed by the <see cref="Signer"/>.
+        /// </summary>
+        /// <param name="publicKey">A public key of the <see cref="Signer"/>.</param>
+        /// <returns><c>true</c> if and only if the given <paramref name="publicKey"/> is
+        /// <see cref="Signer"/>'s and the <see cref="Signature"/> is certainly signed by
+        /// the <see cref="Signer"/>.</returns>
+        [Pure]
+        public bool Verify(PublicKey publicKey) =>
+            Signer.Equals(new Address(publicKey)) &&
+            publicKey.Verify(GetMessage(), Signature.ToBuilder().ToArray());
```

Comment:
Is `ToBuilder()` needed?

Reply:
Without `ToBuild()` the array is built through LINQ which is way general, and I thought building an array through `ImmutableArray<T>.Builder`, which is specific and aware of the implementation details of `ImmutableArray<T>`, would be more efficient, but it's gut feeling. 🙄